README.md
# generator-zf5

[Yeoman](http://yeoman.io) generator for [Zurb Foundation 5](http://foundation.zurb.com/).

### Learn more about it (blog post):
[Foundation 5 and Yeoman generator-zf5](http://julian.io/foundation-5-and-yeoman/)

### Important note:

**Default option here will be Sass with Libsass (for now). But you can choose Ruby version on startup.**

[![NPM](https://nodei.co/npm/generator-zf5.png?downloads=true)](https://nodei.co/npm/generator-zf5/)

[![NPM](https://nodei.co/npm-dl/generator-zf5.png?height=2)](https://nodei.co/npm/generator-zf5/)

## Yo Foundation 5!
* Sass compiling
* autoprefixer (app.css file)
* Publishing to dist directory
* Server with LiveReload (127.0.0.1:9000)
* Bower install
* JSHint
* Custom Modernizr build based on your features usage
* Font Awesome (option)

## Getting Started

```
$ npm install -g yo
```

To install generator-zf5 from npm, run:

```
$ npm install -g generator-zf5
```

Finally, initiate the generator:

```
$ yo zf5
```

## Grunt tasks:

run project
(compile Sass, bower install, livereload (server on 127.0.0.1:9000), watch)
```
$ grunt
```
publishing project (into dist directory)
(compile Sass, validate-js, copy, concatenation, minifications)
```
$ grunt publish
```
dist directory preview (server on 127.0.0.1:9001)
```
$ grunt server-dist
```

### Other Grunt tasks (if you want to use it)

..for validating javascript
```
$ grunt validate-js
```
..for injecting bower libraries (also in default grunt task)
```
$ grunt bower-install
```
..for compiling Sass files + autoprefixer on app.css file
```
$ grunt compile-sass
```

### Ruby Sass with Compass or Node Sass (Libsass)

From version 0.7.0 you can use Ruby version of Sass with Compass. If you want to use Ruby version first of all you need to install compass by 'gem install compass' (it will install Sass gem too).

You don't need to use the config.rb file, all is configured in Gruntfile.js (Sass block). There will be also Compass imports in _appstyles.scss (You can modify it). If you have any problems with using Sass with Compass it is good to uninstall any of your Sass gems and Compass gems and install only Compass gem again. It will fetch proper version of Sass gem.

Ruby Sass config info: <a href="https://github.com/gruntjs/grunt-contrib-sass">https://github.com/gruntjs/grunt-contrib-sass</a>

You can also use Libsass version (default) which is very fast.

Node Sass config info: <a href="https://github.com/sindresorhus/grunt-sass">https://github.com/sindresorhus/grunt-sass</a>

Please test this and send issues if any.

### LiveReload

For LiveReload call 'grunt' (watching) command and go to http://127.0.0.1:9000

### Usemin

Read more about [grunt-usemin](https://github.com/yeoman/grunt-usemin)

### Bower-install

Now you can install your libraries much faster. Example:
```
bower search magnific-popup
...
bower install magnific-popup --save
...
grunt bower-install
```
This should inject the proper js and css paths into your html files. But you should be careful and check what was injected.
'grunt publish' will then minify and concatenate them into a clean (libraries.min.css and libraries.min.js) files.
Instead of a 'bower install' with '--save' you can manualy edit the bower.json file and then run a 'grunt bower-install'. It is also included in the default task - 'grunt'.

## Tips

- Sometimes after new version is released if you have errors when running ````yo zf5```` You should run ````npm cache clean````
- If you have problems with permissions in Linux run this : ````sudo chown -R `whoami` ~/.npm````
- if you want you can delete not used javascript components in index.html file. All remaining components will be minified and concatenated into one foundation.min.js
- if you have problems with connection to http://127.0.0.1:9000 change 'hostname' in Gruntfile.js 'connect' config. Just add ```hostname: '[your hostname]'``` line to ```options: {...}```
- if you want you can delete unnecessary/unused Foundation components from main app.scss (it will be lightest main Foundation css file)
- place all your html files in the root folder (app) or you have to change assets paths (build etc.)
- grunt useminPrepare reference file is only index.html (prevents multiple the same operations) but all html files will be processed, so remember to keep the same usemin 'comments blocks' in all your html files (for now it is good to simply copy index.html, rename it and leave header and footer css and js inclusions with 'comments blocks')
- try to avoid situation when you have the same build blocks in two html files with different assets so (examples):

```
<!-- build:js js/mfpopup/mfpopup.min.js -->
    <script src="js/mfpopup/mfpopup.js"></script>
<!-- endbuild -->
```
and
```
<!-- build:css css/mfpopup/mfpopup.min.js -->
    <script src="js/mfpopup/mfpopup.js"></script>
    <script src="js/mfpopup/other_script.js"></script>
<!-- endbuild -->
```
you can add new ones

- always verify what 'grunt bower-install' injects
- You must look aut where you initialize your project. It is better to not initialize your projec in a subfolder next to .yo-rc.json because your files will land here and not in your subfolder from where you are initializing project
- if you use Compass.. place your mixin includes after Foundation scss partials - [more info](https://github.com/juliancwirko/generator-zf5/issues/18) (by default they are in _appstyles.scss)

You can test it and tell me please if something is not working.

### Getting To Know Yeoman

Yeoman has a heart of gold. He's a person with feelings and opinions, but he's very easy to work with. If you think he's too opinionated, he can be easily convinced.

If you'd like to get to know Yeoman better and meet some of his friends, [Grunt](http://gruntjs.com) and [Bower](http://bower.io), check out the complete [Getting Started Guide](https://github.com/yeoman/yeoman/wiki/Getting-Started).

### License

[MIT License](http://en.wikipedia.org/wiki/MIT_License)

### Contact

[@juliancwirko](https://twitter.com/JulianCwirko) | [julian.cwirko@gmail.com](mailto:julian.cwirko@gmail.com)

### Changelog

..see [CHANGELOG.md](https://github.com/juliancwirko/generator-zf5/blob/master/CHANGELOG.md) file


CHANGELOG.md
#### 1.0.0 (07.10.2015)

- Jade was removed
- grunt-modernizr was removed
- autoprefixer was added
- Foundation was updated to newest v5.5.3

#### 0.10.0 (04.10.2015)

- Modernizr fix - now there is a custom build based on [grunt-modernizr](https://github.com/modernizr/grunt-modernizr) task

#### 0.9.11 (03.10.2015)

- updates
- merge fix

#### 0.9.10 (15.08.2015)

- updates

#### 0.9.9 (15.07.2015)

- versions bump

#### 0.9.8 (01.05.2015)

- Foundation 5.5.2 adjustments
- versions bump

#### 0.9.7 (21.02.2015)

- node-sass 2.0.1
- grunt-sass 0.18.0
- other versions bump

#### 0.9.6 (04.02.2015)

- Foundation 5.5.1 adjustments

#### 0.9.5 (13.12.2014)

- Foundation 5.5.0 adjustments

#### 0.9.4 (11.12.2014)

- Foundation Kitchen Sink added
- bump deps

#### 0.9.3 (28.10.2014)

- Foundation 5.4.7 adjustments
- Foundation works with Libsass 3.0 and Compass 1.0.1 + Sass 3.3 (!!!)

#### 0.9.2 (10.10.2014)

- Foundation 5.4.6 adjustments
- Revert to Libsass version as default

#### 0.9.1 (26.09.2014)

- Foundation 5.4.5 adjustments

#### 0.9.0 (16.09.2014)

- Foundation 5.4.4 adjustments
- Ruby version as default now
- bump deps

#### 0.8.5 (30.08.2014)

- Foundation 5.4.3 adjustments
- added Compass imports in _appstyles.scss
- FontAwesome 4.2.0
- bump deps

#### 0.8.4 (23.08.2014)

- Foundation 5.4.0 adjustments

#### 0.8.3 (31.07.2014)

- Foundation 5.3.2 - 5.3.3 adjustments

#### 0.8.2 (19.07.2014)

- Foundation 5.3.1 adjustments

#### 0.8.1 (11.07.2014)

- Merge pull request from chrisjlee (Remove loading of all gruntfiles; add 'load-grunt-task' dependancy)

#### 0.8.0 (21.06.2014)

- Jade templating engine (option)

#### 0.7.3 (21.06.2014)

- Foundation 5.3.0 adjustments

#### 0.7.2 (31.05.2014)

- Foundation 5.2.3 adjustments

#### 0.7.1 (23.05.2014)

- Font Awesome update
- Deps update

#### 0.7.0 (02.05.2014)

- Ruby Sass with Compass as an option
- Imagemin

#### 0.6.4 (04.04.2014)

- Foundation 5.2.2 adjustments

#### 0.6.3 (21.03.2014)

- grunt useminPrepare - based only on one file - index.html (prevents multiple the same operations if there are many html files)

#### 0.6.2 (20.03.2014)

- Foundation JavaScript file - separation into components
- 'connect' hostname - back to 127.0.0.1

#### 0.6.1 (14.03.2014)

- Foundation 5.2.1 adjustments

#### 0.6.0 (8.03.2014)

- yeoman-generator 0.16.0 cleanup [http://yeoman.io/blog/cleanup.html]
- template folder option removed
- grunt-bower-install added
- Foundation 5.2 cleanup


app/index.js
'use strict';
var util = require('util');
var path = require('path');
var yeoman = require('yeoman-generator');
var chalk = require('chalk');
var mkdirp = require('mkdirp');

var Zf5Generator = yeoman.generators.Base.extend({

	init: function () {
		this.pkg = require('../package.json');
		this.config.save();

		this.on('end', function () {
			if (!this.options['skip-install']) {
				this.installDependencies();
			}
		});
	},

	askForFontAwesome: function () {
		var cb = this.async();

		// have Yeoman greet the user.
		this.log(this.yeoman);

		this.log(chalk.bold.blue('=================='));
		this.log(chalk.bold.blue(' Yo Foundation 5!'));
		this.log(chalk.bold.blue('=================='));

		var prompts = {
			type: 'confirm',
			name: 'fontAwesome',
			message: 'Would you like to include Font Awesome? (Font Awesome gives you scalable vector icons..)',
			default: true
		};

		this.prompt(prompts, function (props) {
			this.fontAwesome = props.fontAwesome;

			cb();
		}.bind(this));
	},

	askForCompass: function () {
		var cb = this.async();

		var prompts = {
			type: 'confirm',
			name: 'compass',
			message: 'Would you like to use Scss with Compass? (default: Scss with Libsass)',
			default: false
		};

		this.prompt(prompts, function (props) {
			this.compass = props.compass;

			cb();
		}.bind(this));
	},

	app: function () {
		mkdirp('dist');
		mkdirp('app');
		this.template('bower.json', 'bower.json');
		this.template('package.json', 'package.json');
		this.template('Gruntfile.js', 'Gruntfile.js');
		this.copy('.jshintrc', '.jshintrc');
		this.copy('.bowerrc', '.bowerrc');
		this.copy('gitignore', '.gitignore');
		this.copy('README.md', 'README.md');
		this.template('index.html', 'app/index.html');
		mkdirp('app/fonts');
		mkdirp('app/images');
		mkdirp('app/js');
		mkdirp('app/css');
		mkdirp('app/scss');
		this.copy('scss/app.scss', 'app/scss/app.scss');
		this.copy('scss/_settings.scss', 'app/scss/_settings.scss');
		this.template('scss/_appstyles.scss', 'app/scss/_appstyles.scss');
		this.copy('js/app.js', 'app/js/app.js');
		this.copy('css/template_override.css', 'app/css/app_override.css');
	},

    install: function () {
        this.installDependencies();
    }

});

module.exports = Zf5Generator;


app/templates/Gruntfile_proj.js
'use strict';

module.exports = function(grunt) {
	grunt.initConfig({
		pkg: grunt.file.readJSON('package.json'),

		sass: {
			options: {
				includePaths: ['bower_components/foundation/scss']
			},
			dist: {
				options: {
					outputStyle: 'extended'
				},
				files: {
					'app/<%= CommonDirName %>/css/app.css': 'app/<%= CommonDirName %>/scss/app.scss',
					'app/<%= ProjectName %>_template/css/<%= ProjectName %>.css': 'app/<%= ProjectName %>_template/scss/<%= ProjectName %>.scss'
				}
			}
		},

		jshint: {
			options: {
				jshintrc: '.jshintrc'
			},
			all: [
				'Gruntfile.js',
				'app/<%= ProjectName %>_template/js/{,*/}*.js'
			]
		},

		clean: {
			dist: {
				src: ['dist/*']
			},
		},

		copy: {
			dist: {
				files: [{
					expand: true,
					cwd:'app/',
					src: ['*.html'],
					dest: 'dist/',
					filter: 'isFile'
				}, {
					expand: true,
					cwd:'app/<%= CommonDirName %>',
					src: ['css/**', 'js/**', 'images/**', 'fonts/**', '!**/*.scss'],
					dest: 'dist/<%= CommonDirName %>'
				}, {
					expand: true,
					cwd:'app/<%= ProjectName %>_template',
					src: ['css/**', 'js/**', 'images/**', 'fonts/**', '!**/*.scss'],
					dest: 'dist/<%= ProjectName %>_template'
				}, {
					expand: true,
					flatten: true,
					src: ['bower_components/jquery/jquery.min.js', 'bower_components/modernizr/modernizr.js'],
					dest: 'dist/<%= CommonDirName %>/js/vendor/',
					filter: 'isFile'
				}, {
					expand: true,
					flatten: true,
					src: ['bower_components/foundation/js/foundation.min.js'],
					dest: 'dist/<%= CommonDirName %>/js/foundation/',
					filter: 'isFile'
				}<% if (fontAwesome) { %> , {
					expand: true,
					flatten: true,
					src: ['bower_components/font-awesome/fonts/**'],
					dest: 'dist/<%= CommonDirName %>/fonts/',
					filter: 'isFile'
				},
				{
					expand: true,
					flatten: true,
					src: ['bower_components/font-awesome/css/font-awesome.min.css'],
					dest: 'dist/<%= CommonDirName %>/css/',
					filter: 'isFile'
				}<% } %>]
			},
		},

		useminPrepare: {
			html: 'app/*.html',
			options: {
				dest: 'dist'
			}
		},

		usemin: {
			html: ['dist/*.html'],
			css: ['dist/<%= ProjectName %>_template/css/*.css', 'dist/<%= CommonDirName %>/css/*.css'],
			options: {
				dirs: ['dist']
			}
		},

		watch: {
			grunt: {
				files: ['Gruntfile.js']
			},
			sass: {
				files: ['app/<%= CommonDirName %>/scss/{,*/}*.scss', 'app/<%= ProjectName %>_template/scss/{,*/}*.scss'],
				tasks: ['sass']
			},
			livereload: {
				files: ['app/*.html', 'app/<%= CommonDirName %>/js/{,*/}*.js', 'app/<%= ProjectName %>_template/js/{,*/}*.js', 'app/<%= CommonDirName %>/css/{,*/}*.css', 'app/<%= ProjectName %>_template/css/{,*/}*.css', 'app/<%= CommonDirName %>/images/{,*/}*.{jpg,gif,svg,jpeg,png}', 'app/<%= ProjectName %>_template/images/{,*/}*.{jpg,gif,svg,jpeg,png}'],
				options: {
					livereload: true
				}
			}
		}

	});

	grunt.loadNpmTasks('grunt-sass');
	grunt.loadNpmTasks('grunt-contrib-watch');
	grunt.loadNpmTasks('grunt-contrib-clean');
	grunt.loadNpmTasks('grunt-contrib-copy');
	grunt.loadNpmTasks('grunt-contrib-jshint');
	grunt.loadNpmTasks('grunt-usemin');

	grunt.registerTask('build', ['sass']);
	grunt.registerTask('default', ['build', 'watch']);
	grunt.registerTask('validate-js', ['jshint']);
	grunt.registerTask('publish', ['clean:dist', 'validate-js', 'useminPrepare', 'copy:dist', 'usemin']);

};

app/templates/Gruntfile.js
'use strict';

module.exports = function(grunt) {

    require('load-grunt-tasks')(grunt);


    grunt.initConfig({
        pkg: grunt.file.readJSON('package.json'),
        app: 'app',
        dist: 'dist',

        sass: {<% if (!compass) { %>
            options: {
                includePaths: ['<%%= app %>/bower_components/foundation/scss']
            },
            dist: {
                options: {
                    outputStyle: 'extended'
                },
                files: {
                    '<%%= app %>/css/app.css': '<%%= app %>/scss/app.scss'
                }
            }<% } else { %>
            dist: {
                options: {
                    style: 'expanded', // expanded or nested or compact or compressed
                    loadPath: '<%%= app %>/bower_components/foundation/scss',
                    compass: true,
                    quiet: true
                },
                files: {
                    '<%%= app %>/css/app.css': '<%%= app %>/scss/app.scss'
                }
            }<% } %>
        },

        postcss: {
            options: {
                processors: [
                    require('autoprefixer')({browsers: 'last 2 versions'})
                ]
            },
            dist: {
                src: '<%%= app %>/css/app.css'
            }
        },

        jshint: {
            options: {
                jshintrc: '.jshintrc'
            },
            all: [
                'Gruntfile.js',
                '<%%= app %>/js/**/*.js'
            ]
        },

        clean: {
            dist: {
                src: ['<%%= dist %>/*']
            },
        },
        copy: {
            dist: {
                files: [{
                    expand: true,
                    cwd:'<%%= app %>/',
                    src: ['fonts/**', '**/*.html', '!**/*.scss', '!bower_components/**'],
                    dest: '<%%= dist %>/'
                }<% if (fontAwesome) { %> , {
                    expand: true,
                    flatten: true,
                    src: ['<%%= app %>/bower_components/font-awesome/fonts/**'],
                    dest: '<%%= dist %>/fonts/',
                    filter: 'isFile'
                } <% } %>]
            },
        },

        imagemin: {
            target: {
                files: [{
                    expand: true,
                    cwd: '<%%= app %>/images/',
                    src: ['**/*.{jpg,gif,svg,jpeg,png}'],
                    dest: '<%%= dist %>/images/'
                }]
            }
        },

        uglify: {
            options: {
                preserveComments: 'some',
                mangle: false
            }
        },

        useminPrepare: {
            html: ['<%%= app %>/index.html'],
            options: {
                dest: '<%%= dist %>'
            }
        },

        usemin: {
            html: ['<%%= dist %>/**/*.html', '!<%%= app %>/bower_components/**'],
            css: ['<%%= dist %>/css/**/*.css'],
            options: {
                dirs: ['<%%= dist %>']
            }
        },

        watch: {
            grunt: {
                files: ['Gruntfile.js'],
                tasks: ['sass', 'postcss']
            },
            sass: {
                files: '<%%= app %>/scss/**/*.scss',
                tasks: ['sass', 'postcss']
            },
            livereload: {
                files: ['<%%= app %>/**/*.html', '!<%%= app %>/bower_components/**', '<%%= app %>/js/**/*.js', '<%%= app %>/css/**/*.css', '<%%= app %>/images/**/*.{jpg,gif,svg,jpeg,png}'],
                options: {
                    livereload: true
                }
            }
        },

        connect: {
            app: {
                options: {
                    port: 9000,
                    base: '<%%= app %>/',
                    open: true,
                    livereload: true,
                    hostname: '127.0.0.1'
                }
            },
            dist: {
                options: {
                    port: 9001,
                    base: '<%%= dist %>/',
                    open: true,
                    keepalive: true,
                    livereload: false,
                    hostname: '127.0.0.1'
                }
            }
        },

        wiredep: {
            target: {
                src: [
                    '<%%= app %>/**/*.html'
                ],
                exclude: [
                    'modernizr',<% if (fontAwesome) { %>
                    'font-awesome',<% } %>
                    'jquery-placeholder',
                    'foundation'
                ]
            }
        }

    });

    grunt.registerTask('compile-sass', ['sass', 'postcss']);
    grunt.registerTask('bower-install', ['wiredep']);
    grunt.registerTask('default', ['compile-sass', 'bower-install', 'connect:app', 'watch']);
    grunt.registerTask('validate-js', ['jshint']);
    grunt.registerTask('server-dist', ['connect:dist']);
    grunt.registerTask('publish', ['compile-sass', 'clean:dist', 'validate-js', 'useminPrepare', 'copy:dist', 'newer:imagemin', 'concat', 'cssmin', 'uglify', 'usemin']);

};
